mllib/src/main/scala/org/apache/spark/ml/reduction/OneVsAll.scala

```diff
@@ -0,0 +1,165 @@
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema, fitting = false, featuresDataType)
+  }
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    // Check schema
+    val parentSchema = dataset.schema
+    transformSchema(parentSchema, logging = true)
+    val sqlCtx = dataset.sqlContext
+
+    // score each model on every data point and pick the model with highest score
+    // TODO: Use DataFrame expressions to leverage performance here
+    val predictions = models.zipWithIndex.par.map { case (model, index) =>
+      val output = model.transform(dataset)
+      output.select($(rawPredictionCol)).map { case Row(p: Vector) => List((index, p(1))) }
+    }.reduce[RDD[List[(Int, Double)]]] { case (x, y) =>
+      x.zip(y).map { case ((a, b)) =>
+        a ++ b
+      }
+    }.map(_.maxBy(_._2))
+
+    // ensure that we pass through columns that are part of the original dataset.
+    val results = dataset.select(col("*")).rdd.zip(predictions).map { case (row, (label, _)) =>
+      Row.fromSeq(row.toSeq ++ List(label.toDouble))
+    }
+
+    // the output schema should retain all input fields and add prediction column.
+    val outputSchema = SchemaUtils.appendColumn(parentSchema, $(predictionCol), DoubleType)
+    sqlCtx.createDataFrame(results, outputSchema)
+  }
+
+  def predict(features: FeaturesType): Double = {
+    throw new UnsupportedOperationException("Ensemble Classifier does not support predict," +
+      " use transform instead")
+  }
+}
+
+/**
+ * :: Experimental ::
+ *
+ * Reduction of Multiclass Classification to Binary Classification.
+ * Performs reduction using one against all strategy.
+ * For a multiclass classification with k classes, train k models (one per class).
+ * Each example is scored against all k models and the model with highest score
+ * is picked to label the example.
+ *
+ * Classifier parameters like featuresCol, predictionCol and rawPredictionCol are
+ * set directly on the OneVsRest and are ignored on the underlying classifier.
+ *
+ */
+@Experimental
+class OneVsAll[
+    FeaturesType,
+    E <: Classifier[FeaturesType, E, M],
+    M <: ClassificationModel[FeaturesType, M]]
+  (val classifier: Classifier[FeaturesType, E, M])
+  (implicit m: ClassTag[M])
+  extends Predictor[FeaturesType, OneVsAll[FeaturesType, E, M], OneVsAllModel[FeaturesType, E, M]]
+  with ClassifierParams {
+
+  override private[ml] def featuresDataType: DataType = classifier.featuresDataType
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema, fitting = true, featuresDataType)
+  }
+
+  override protected def train(dataset: DataFrame): OneVsAllModel[FeaturesType, E, M] = {
+    // determine number of classes either from metadata if provided, or via computation.
+    val labelSchema = dataset.schema($(labelCol))
+    val computeNumClasses: () => Int = () => {
+      dataset.select($(labelCol)).distinct.count().toInt
```

Comment:
This should search for the maximum value, not the number of distinct ones.  Later on, there's the assumption that classes use indices 0,...,numClasses-1